sklearn/tree/_tree.pyx

```diff
@@ -1133,14 +1115,14 @@ cdef class Tree:
         shape[0] = <cnp.npy_intp> self.node_count
         shape[1] = <cnp.npy_intp> self.n_outputs
         shape[2] = <cnp.npy_intp> self.max_n_classes
-        cdef cnp.ndarray arr
+        cdef DOUBLE_t[:, :, ::1] arr
         arr = cnp.PyArray_SimpleNewFromData(3, shape, cnp.NPY_DOUBLE, self.value)
         Py_INCREF(self)
-        if PyArray_SetBaseObject(arr, <PyObject*> self) < 0:
+        if PyArray_SetBaseObject(arr.base, <PyObject*> self) < 0:
             raise ValueError("Can't initialize array.")
         return arr
 
-    cdef cnp.ndarray _get_node_ndarray(self):
+    cdef Node[::1] _get_node_ndarray(self):
```

Comment:
Given the naming of `_get_node_ndarray`, I prefer it to return a `ndarray` instead of a memoryview.

Reply:
In that case the return type should be defined as `np.ndarray` right?

Reply:
Likely needs to be `cnp.ndarray` because the signature is in `_tree.pxd`, which requires `cimport cnp`.

Overall, I prefer not to change the interface for `_apply_dense`, `_apply_sparse_csr`, `_get_value_ndarray` and `_get_node_ndarray`. As shown in this PR, it forces the caller to call `np.asarray` on the output.

Reply:
Then it is basically fine to define the return type as a cnp.ndarray? Even though we are replacing the deprecated cnp.ndarray with memory views.

Reply:
From my understanding `cnp.ndarray` itself was not deprecated. Directly accessing `cnp.ndarray`'s attributes such as `.data` was deprecated. In this case, returning `cnp.ndarray` should still work even with `NPY_1_7_API_VERSION`. (To confirm, I locally tested returning `cnp.ndarray` with your PR and it worked.)

Reply:
Yeah I think using cnp.ndarray as a return type should be fine. I also tried it out.

Reply:
Thank you for the suggestion.